prune/prune.go

```
@@ -17,181 +17,197 @@ package prune
 }
 
-// StrategyFunc function allows a means to specify
-// custom prune strategies
-type StrategyFunc func(ctx context.Context, cfg Config, resources []ResourceInfo) ([]ResourceInfo, error)
-
-// PreDelete function is called before a resource is pruned
-type PreDelete func(ctx context.Context, cfg Config, something ResourceInfo) error
-
-// Config defines a pruning configuration and ultimately
-// determines what will get pruned
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// Question: Where should this be used? Why should this be used over the Go default error?
+// ErrUnpruneable indicates that it is not allowed to prune a specific object.
+type ErrUnpruneable struct {
+	Obj    *client.Object
+	Reason string
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
+
+// IsPruneableFunc is a function that checks a the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `ErrUnpruneable` if it is unsafe, or another error.
```

I don't see anything returning `ErrUnpruneable`. Either the methods should return this error or we should update the comment to reflect what is really being returned.